### PR TITLE
UCP/WORKER_FENCE: added worker_fence over EPs

### PR DESCRIPTION
**Pull request by hoopoepg**
- added implementation of worker_fence via EPs flush_remote calls
- added test

### CODE REVIEW

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -314,6 +314,12 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "another thread, or incoming active messages, but consumes more resources.",
    ucs_offsetof(ucp_context_config_t, flush_worker_eps), UCS_CONFIG_TYPE_BOOL},
 
+  {"FENCE_WORKER_EPS", "auto",
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:317]: maybe better use `WORKER_STRONG_FENCE`

File: src/ucp/core/ucp_context.c [hunk 2]
```diff
@@ -314,6 +314,12 @@ static ucs_config_field_t ucp_context_config_table[] = {
...
+   "Enable flushing the worker by fencing its endpoints. Allows completing\n"
+   "the fence operation in a bounded time even if there are new requests on\n"
+   "another thread, or incoming active messages, but consumes more resources.",
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:320]: it does not affect flush, i'd rephrase it, like:
`Enable strong sync mode for ucp_worker_fence. By default it is used when multiple lanes are used for RMA operations.`

File: src/ucp/rma/flush.c [hunk 3]
```diff
@@ -383,7 +383,8 @@ ucs_status_ptr_t ucp_ep_flush_internal(ucp_ep_h ep, unsigned req_flags,
     req->send.ep                    = ep;
     req->send.flushed_cb            = flushed_cb;
     req->send.flush.prog_id         = UCS_CALLBACKQ_ID_NULL;
-    req->send.flush.uct_flags       = UCT_FLUSH_FLAG_LOCAL;
+    req->send.flush.uct_flags       = ucp_request_param_flags(param,
+                                                              UCT_FLUSH_FLAG_LOCAL);
```
Comment (brminich) [inline src/ucp/rma/flush.c:387]: can we take it from `worker_req` instead? It does not seem to be correct to pass UCT flags thru `UCP_request_param_t`

File: src/ucp/rma/flush.c [hunk 4]
```diff
@@ -680,24 +688,56 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
...
+            }
+        }
+        status = UCS_OK;
+    } else {
+        /* Suppress false positive cpp-checker hits */
+        request.flush_worker.comp_count = 0;
+        request.status                  = UCS_OK;
+        /* force using flush on EPs */
+        ucp_worker_flush_ops_count_add(worker, 1);
+        status_ptr = ucp_worker_flush_nbx_internal(worker, &param);
```
Comment (brminich) [inline src/ucp/rma/flush.c:724]: will it be difficult to implement ignoring those ep which have just 1 `rma_bw_lane` configured? I think for the fence we do not need to do anything with them

File: src/ucp/rma/flush.c [hunk 5]
```diff
@@ -680,24 +688,56 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
 
 UCS_PROFILE_FUNC(ucs_status_t, ucp_worker_fence, (worker), ucp_worker_h worker)
 {
+    ucp_context_t *context          = worker->context;
+    ucp_request_t request;
+    const ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS | UCP_OP_ATTR_FIELD_REQUEST,
+        .flags        = UCT_FLUSH_FLAG_REMOTE,
+        .request      = UCS_PTR_BYTE_OFFSET(&request, sizeof(request))
```
Comment (brminich) [inline src/ucp/rma/flush.c:696]: why do we need request here?

File: src/ucp/core/ucp_context.c [hunk 6]
```diff
@@ -314,6 +314,11 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "another thread, or incoming active messages, but consumes more resources.",
    ucs_offsetof(ucp_context_config_t, flush_worker_eps), UCS_CONFIG_TYPE_BOOL},
 
+  {"WORKER_STRONG_FENCE", "auto",
+   "Enable strong sync mode for ucp_worker_fence. By default it is used when\n"
+    "multiple lanes are used for RMA operations.",
+   ucs_offsetof(ucp_context_config_t, fence_worker_eps), UCS_CONFIG_TYPE_ON_OFF_AUTO},
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c:320]: renamed

File: src/ucp/rma/flush.c [hunk 7]
```diff
@@ -680,24 +685,56 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
 
 UCS_PROFILE_FUNC(ucs_status_t, ucp_worker_fence, (worker), ucp_worker_h worker)
 {
+    ucp_context_t *context          = worker->context;
+    ucp_request_t request;
+    const ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS | UCP_OP_ATTR_FIELD_REQUEST,
+        .flags        = UCT_FLUSH_FLAG_REMOTE,
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/rma/flush.c]: ok, removed

File: src/ucp/core/ucp_context.c [hunk 8]
```diff
@@ -1843,7 +1843,7 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
     memcpy(context->config.am_mpools.sizes, config->mpool_sizes.memunits,
            config->mpool_sizes.count * sizeof(size_t));
 
-    context->config.worker_fence_eps =
+    context->config.worker_strong_fence =
             (context->config.ext.fence_worker_eps == UCS_CONFIG_ON) ||
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: yep, missed

File: src/ucp/rma/flush.c [hunk 9]
```diff
@@ -680,24 +686,56 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
 
 UCS_PROFILE_FUNC(ucs_status_t, ucp_worker_fence, (worker), ucp_worker_h worker)
 {
+    ucp_context_t *context          = worker->context;
+    ucp_request_t request;
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/rma/flush.c]: removed

**Review by brminich — DISMISSED** [review]

File: src/ucp/core/ucp_context.c [hunk 10]
```diff
@@ -314,6 +314,11 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "another thread, or incoming active messages, but consumes more resources.",
    ucs_offsetof(ucp_context_config_t, flush_worker_eps), UCS_CONFIG_TYPE_BOOL},
 
+  {"WORKER_STRONG_FENCE", "auto",
+   "Enable strong sync mode for ucp_worker_fence. By default it is used when\n"
+    "multiple lanes are used for RMA operations.",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:319]: 1. indent
2. Maybe it could be be enum "UCX_FENCE_MODE=[weak|strong|auto]" so we could add more options in the future?

File: src/ucp/rma/flush.c [hunk 11]
```diff
@@ -558,7 +560,8 @@ static unsigned ucp_worker_flush_progress(void *arg)
         }
     }
 
-    if (worker->context->config.ext.flush_worker_eps &&
+    if ((worker->context->config.ext.flush_worker_eps ||
+         (req->flush_worker.uct_flags == UCT_FLUSH_FLAG_REMOTE)) &&
```
Comment (yosefe) [inline src/ucp/rma/flush.c:564]: maybe & instead of == ?

File: src/ucp/rma/flush.c [hunk 12]
```diff
@@ -588,7 +591,7 @@ static unsigned ucp_worker_flush_progress(void *arg)
 
 static ucs_status_ptr_t
 ucp_worker_flush_nbx_internal(ucp_worker_h worker,
-                              const ucp_request_param_t *param)
+                              const ucp_request_param_t *param, unsigned flags)
```
Comment (yosefe) [inline src/ucp/rma/flush.c:594]: flags -> uct_flags

File: src/ucp/rma/flush.c [hunk 13]
```diff
@@ -680,24 +686,51 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
...
     ucs_status_t status;
 
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
-    UCS_BITMAP_FOR_EACH_BIT(worker->context->tl_bitmap, rsc_index) {
-        wiface = ucp_worker_iface(worker, rsc_index);
-        if (wiface->iface == NULL) {
-            continue;
-        }
+    if (!context->config.worker_strong_fence) {
```
Comment (yosefe) [inline src/ucp/rma/flush.c:698]: 1. switch if/else
2. maybe separate to 2 helper functions? since they're doing 2 completely different things

File: test/gtest/ucp/test_ucp_fence.cc [hunk 14]
```diff
@@ -113,10 +124,15 @@ class test_ucp_fence : public ucp_test {
                      ucp_rkey_h rkey, void *memheap_ptr,
                      uint64_t initial_value, uint32_t* error) {
         ucs::ptr_vector<worker> m_workers;
+        std::vector<entity*> entities;
         m_workers.clear();
         m_workers.push_back(new worker(this, send1, send2, sender, rkey,
                                        memheap_ptr, initial_value, error));
-        m_workers.at(0).join();
+        entities.push_back(&receiver());
+        while (m_workers.at(0).join_np()) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_fence.cc:132]: 1. can we count running threads in the test class (atomic var for example) instead of using pthread_tryjoin_np?
2. maybe use pthread barrier to sync threads before they exit?

Comment (hoopoepg) [reply test/gtest/ucp/test_ucp_fence.cc]: re-designed test

File: src/ucp/rma/flush.c [hunk 15]
```diff
@@ -684,55 +685,67 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
...
-        do {
-            ucp_worker_progress(worker);
-        } while (request->flush_worker.comp_count > 1);
-        ucp_worker_flush_complete_one(request, request->status, 1);
-        status = request->status;
-        ucp_request_put(request);
-out_ops_count:
+    if (worker->context->config.worker_strong_fence) {
+        /* force using flush on EPs */
+        ucp_worker_flush_ops_count_add(worker, 1);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:742]: IMO need to move ucp_worker_flush_ops_count_add inside ucp_worker_fence_strong

Comment (hoopoepg) [reply src/ucp/rma/flush.c]: it makes code bit more complex... ok, done

**Review by yosefe — DISMISSED** [review]

**Comment by yosefe** [comment]
@hoopoepg pls squash

**Review by yosefe — DISMISSED** [review]

File: src/ucp/core/ucp_context.c [hunk 16]
```diff
@@ -314,6 +321,14 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "another thread, or incoming active messages, but consumes more resources.",
    ucs_offsetof(ucp_context_config_t, flush_worker_eps), UCS_CONFIG_TYPE_BOOL},
 
+  {"FENCE_MODE", "auto",
+   "Fence mode.\n"
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: ok, added reference to ucp_worker_fence routine

**Review by brminich — DISMISSED** [review]

**Comment by hoopoepg** [comment]
@yosefe ok to squash again?

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]
